src/sqlfluff/rules/L050.py

```diff
@@ -0,0 +1,122 @@
+
+
+@document_fix_compatible
+class Rule_L050(BaseRule):
+    """Files must not begin with newlines or whitespace.
+
+    | **Anti-pattern**
+    | The content in file begins with newlines or whitespace, the ^ represents the beginning of file.
+
+    .. code-block:: sql
+       :force:
+
+        ^
+
+        SELECT
+            a
+        FROM foo
+
+        -- Beginning on an indented line is also forbidden,
+        -- (the • represents space).
+
+        ••••SELECT
+        ••••a
+        FROM
+        ••••foo
+
+    | **Best practice**
+    | Start file on either code or comment, the ^ represents the beginning of file.
+
+    .. code-block:: sql
+       :force:
+
+
+        ^SELECT
+            a
+        FROM foo
+
+        -- Including an initial block comment.
+
+        ^/*
+        This is a description of my SQL code.
+        */
+        SELECT
+            a
+        FROM
+            foo
+
+        -- Including an initial inline comment.
+
+        ^--This is a description of my SQL code.
+        SELECT
+            a
+        FROM
+            foo
+    """
+
+    @staticmethod
+    def _potential_template_collision(context: RuleContext) -> bool:
+        """Check for any templated raw slices that intersect with source slices in the raw_stack.
+
+        Returns:
+            :obj:`bool` indicating a preceding templated raw slice has been detected.
+        """
+        templated_file = context.segment.pos_marker.templated_file
+        for segment in context.raw_stack:
+            source_slice = segment.pos_marker.source_slice
+            raw_slices = templated_file.raw_slices_spanning_source_slice(source_slice)
+            if any(
+                raw_slice
+                for raw_slice in raw_slices
+                if raw_slice.slice_type == "templated"
+            ):
+                return True
+
+        return False
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Files must not begin with newlines or whitespace."""
+        # If parent_stack is empty we are currently at FileSegment.
+        if len(context.parent_stack) == 0:
+            return None
+
+        # If raw_stack is empty there can be nothing to remove.
+        if len(context.raw_stack) == 0:
+            return None
+
+        # If the current segment is either comment or code and all
+        # previous segments are forms of whitespace then we can
+        # remove these earlier segments.
+        # Given the tree stucture, we make sure we are at the
+        # first leaf to avoid repeated detection.
+        whitespace_set = {"newline", "whitespace", "Indent", "Dedent"}
+        if (
+            # Non-whitespace segment.
+            (context.segment.name not in whitespace_set)
+            # We want first Non-whitespace segment so
+            # all preceding segments must be whitespace.
+            and set(segment.name for segment in context.raw_stack).issubset(
```

Comment:
Would it be clearer to avoid constructing a temporary set by doing something like this?
```
all(segment.name in whitespace_set for segment in context.raw_stack)
```

Reply:
agree with you on this, making the change.

Reply:
done